File: 0239-sliding-window-maximum/0239-sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        res = []

        window = deque([]) #(index,val)
        for r in range(len(nums)):
            cur = nums[r]
            if window and r-window[0][0]==k:
                window.popleft()
            
            while window and window[-1][1] <= cur:
                window.pop()
            window.append([r,cur])

            if r >= k-1:
                res.append(window[0][1])

        return res

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> res;
        deque<pair<int,int>> window;

        for (int r = 0; r<nums.size(); r++) {
            int cur = nums[r];

            if (!window.empty() && r-window[0].first==k) {
                window.pop_front();
            }

            while (!window.empty() && window.back().second <= cur) {
                window.pop_back();
            }
            window.push_back({r,cur});

            if (r >= k-1){
                res.push_back(window[0].second);
            }
        }
        return res;
    }
};